### PR TITLE
Add ast support module, extracted from rust-analyzer

### PR DESCRIPTION
**Pull request by mullr**
I used this code from rust-analyzer in order to be able to build out my own AST structures. The only change I made was to make it generic over language.

### CODE REVIEW

**Comment by mullr** [comment]
Is there any desire to have this code in the main rowan repo? If not, I can make a separate utility crate. Just don't want to maintain an unnecessary fork.

File: src/ast.rs [hunk 1]
```diff
@@ -162,3 +163,39 @@ impl<N: AstNode> From<AstPtr<N>> for SyntaxNodePtr<N::Language> {
...
+        parent.children().find_map(N::cast)
+    }
+
+    pub fn children<N: AstNode>(parent: &SyntaxNode<N::Language>) -> AstChildren<N> {
+        AstChildren::new(parent)
+    }
+
+    pub fn token<L: Language>(parent: &SyntaxNode<L>, kind: L::Kind) -> Option<SyntaxToken<L>> {
+        parent.children_with_tokens().filter_map(|it| it.into_token()).find(|it| it.kind() == kind)
+    }
```
Comment (Veykril) [inline src/ast.rs:200]: ~~Not entirely sure how useful the `token` will be given you'll have to specify the language type which seems kind of annoying.~~

Actually it should get inferred from the `parent` parameter.

**Review by Veykril — APPROVED** [review]
Sounds reasonable to me to have these exposed for a bit nicer typed handling of things.

**Comment by Veykril** [comment]
Ah, can you bump the rowan version to `0.15.4`? Then merging will auto-publish the new minor version

**Comment by mullr** [comment]
> Ah, can you bump the rowan version to 0.15.4? 

Done

**Pull request merged**